pkg/controller/master/upgrade/upgrade_repo.go

```diff
@@ -454,3 +482,63 @@ func (r *Repo) getInfo() (*RepoInfo, error) {
 	}
 	return &info, nil
 }
+
+func (r *Repo) getImageList(version string, imageList map[string]bool) error {
+	imageListURL := fmt.Sprintf("http://%s.%s/harvester-iso/bundle/harvester/images-lists-archive/%s/image_list_all.txt",
+		r.getRepoServiceName(),
+		upgradeNamespace,
+		version,
+	)
+
+	req, err := http.NewRequestWithContext(r.ctx, http.MethodGet, imageListURL, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status: %s", resp.Status)
+	}
+
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		imageName := scanner.Text()
+		imageList[imageName] = true
+	}
+
+	return scanner.Err()
+}
+
+func (r *Repo) getImagesDiffList() ([]string, error) {
+	previousImageList := make(map[string]bool)
+	currentImageList := make(map[string]bool)
+
+	backoff := wait.Backoff{
+		Steps:    30,
+		Duration: 10 * time.Second,
+		Factor:   1.0,
+		Jitter:   0.1,
+	}
+
+	if err := retry.OnError(backoff, util.IsRetriableNetworkError, func() error {
```

Comment:
Seems the retry mechanism needs some fine-tuning:
```
time="2024-01-25T01:41:07Z" level=info msg="Try to start repo VM for image pruning"
time="2024-01-25T01:41:07Z" level=info msg="Trying to get v1.2.1 image list"
W0125 01:41:08.717537       7 warnings.go:70] kubevirt.io/v1 VirtualMachineInstancePresets is now deprecated and will be removed in v2.
W0125 01:41:08.725561       7 warnings.go:70] v1 ComponentStatus is deprecated in v1.19+
time="2024-01-25T01:41:12Z" level=error msg="Failed to dial steve aggregation server: dial tcp 10.53.242.92:443: connect: connection refused"

time="2024-01-25T01:41:21Z" level=warning msg="Unable to cleanup images: Get \"http://upgrade-repo-hvst-upgrade-qjl4q.harvester-system/harvester-iso/bundle/harvester/images-lists-archive/v1
.2.1/image_list_all.txt\": dial tcp: lookup upgrade-repo-hvst-upgrade-qjl4q.harvester-system on 10.53.0.10:53: server misbehaving"

time="2024-01-25T01:41:21Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-qjl4q with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:Succeeded]"
time="2024-01-25T01:41:21Z" level=info msg="Add OwnerReference upgrade-repo-hvst-upgrade-qjl4q-disk-0 to upgrade repo image harvester-system/image-gg6lg"
time="2024-01-25T01:41:21Z" level=info msg="Delete upgrade repo VM harvester-system/upgrade-repo-hvst-upgrade-qjl4q"
time="2024-01-25T01:41:21Z" level=info msg="Reset RKEConfig and set provisionGeneration to 0"
time="2024-01-25T01:41:21Z" level=info msg="Update upgradeLog harvester-system/hvst-upgrade-qjl4q-upgradelog"
time="2024-01-25T01:41:21Z" level=info msg="Stop collecting logs"
```

Reply:
That seems to be the coredns Pod not ready yet for the Service to redirect the traffic. We need an additional check before the HTTP request is fired as it's not an HTTP error. Something like the following might help:
```
_, err := net.LookupIP("upgrade-repo-hvst-upgrade-qjl4q.harvester-system")
```

Reply:
It makes more sense to classify the DNS-related errors as retriable. No need to implement additional requests and check the results.